### PR TITLE
Projection axis functionality

### PR DESCRIPTION
**Pull request by alexkiri**
This PR adds configurable projection axis functionality
- useful for generating vertical side views
- configurable with argument, the default axis is `z`
- E.g. `bsp2svg.exe ad_tears.bsp -p x`

image
Screenshot of Tears of the False God (ad_tears)

Also includes a fix for the issue with some edges https://imgsli.com/MjE5NjQy
The fix is essentially just line 106 `stroke_miterlimit='0'`

### CODE REVIEW

File: qcli/bsp2svg/converter.py [hunk 1]
```diff
@@ -31,52 +31,67 @@ def convert(bsp_file, svg_file, args):
...
+    projection_axis = args.projection_axis
+    
+    # Filter faces
+    faces = [face for model in bsp_file.models for face in model.faces]
+    ignore_textures = ['clip', 'hint', 'trigger'] + args.ignore
+    faces = list(filter(lambda f: not (f.texture_name.startswith('sky') or f.texture_name in ignore_textures), faces))
 
     # Determine map bounds
-    vs = [vertex[:] for model in bsp_file.models for face in model.faces for vertex in face.vertexes]
+    vs = [vertex[:] for face in faces for vertex in face.vertexes]
```
Comment (joshuaskelly) [inline qcli/bsp2svg/converter.py:42]: Why did you break apart the nested list comprehension? Is it too tricky to understand/reckon about?

File: qcli/bsp2svg/converter.py [hunk 2]
```diff
@@ -31,52 +31,67 @@ def convert(bsp_file, svg_file, args):
...
         svg_file,
-        viewBox=f'{min_x - padding} {min_y - padding} {width + padding * 2} {height + padding * 2}',
+        viewBox=f'{drawing_min_x - padding} {drawing_min_y - padding} {width + padding * 2} {height + padding * 2}',
         profile='tiny'
     )
 
     dwg.add(
         dwg.rect(
-            insert=(min_x - padding, min_y - padding),
+            id='background',
```
Comment (joshuaskelly) [inline qcli/bsp2svg/converter.py:66]: Interesting. What does this id do?

File: qcli/bsp2svg/cli.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-"""Command line utility for creating and creating WAD files from BSP files
+"""Command line utility for creating SVG files from BSP files
```
Comment (joshuaskelly) [inline qcli/bsp2svg/cli.py:1]: Thank you for fixing my bad copy + paste. 👍

File: qcli/bsp2svg/cli.py [hunk 4]
```diff
@@ -71,7 +79,7 @@ def main():
     # Validate or create out file
     if args.dest == os.getcwd():
         svg_path = os.path.dirname(args.file)
-        svg_name = f'{os.path.basename(args.file).split(".")[0]}.svg'
+        svg_name = f'{os.path.basename(args.file).split(".")[0]}_{args.projection_axis}.svg'
```
Comment (joshuaskelly) [inline qcli/bsp2svg/cli.py:82]: I like this!

File: qcli/bsp2svg/converter.py [hunk 5]
```diff
@@ -31,52 +31,67 @@ def convert(bsp_file, svg_file, args):
     """
     print(f'Reading {os.path.basename(bsp_file)}')
     bsp_file = Bsp.open(bsp_file)
+    projection_axis = args.projection_axis
+    
+    # Filter faces
+    faces = [face for model in bsp_file.models for face in model.faces]
+    ignore_textures = ['clip', 'hint', 'trigger'] + args.ignore
+    faces = list(filter(lambda f: not (f.texture_name.startswith('sky') or f.texture_name in ignore_textures), faces))
```
Comment (joshuaskelly) [inline qcli/bsp2svg/converter.py:39]: I don't like the hard coding of the sky texture. That is what the `--ignore` arg is for.

File: qcli/bsp2svg/converter.py [hunk 6]
```diff
@@ -31,52 +31,67 @@ def convert(bsp_file, svg_file, args):
...
         )
     )
 
-    group = dwg.g(
-        id='bsp_ref',
-    )
+    group = dwg.g(id='bsp_ref')
     dwg.defs.add(group)
-
-    ignore_textures = ['clip', 'hint', 'trigger'] + args.ignore
```
Comment (joshuaskelly) [inline qcli/bsp2svg/converter.py:68]: Why did you remove these ignore textures?

File: qcli/bsp2svg/converter.py [hunk 7]
```diff
@@ -31,52 +31,67 @@ def convert(bsp_file, svg_file, args):
...
+    elif projection_axis == 'y':
+        faces.sort(key=lambda f: f.vertexes[0].y)
+    elif projection_axis == 'z':
+        faces.sort(key=lambda f: f.vertexes[0].z)
+    
+    def vs_picker(vertexes):
+        if projection_axis == 'x':
+            return vertexes[1:3]
+        elif projection_axis == 'y':
+            return [vertexes[0], vertexes[2]]
```
Comment (joshuaskelly) [inline qcli/bsp2svg/converter.py]: You don't need to return a list here. You can just return a tuple:
`return vertexes[0], vertexes[2]`

If you think it's more readable, you can do this for all three branches of the if statement instead of slicing.

**Review by joshuaskelly — CHANGES_REQUESTED** [review]
Looks good for a first pass! There are a few things I have questions on + some feedback changes.

File: qcli/bsp2svg/converter.py (hunk 2 above)
Comment (alexkiri) [reply qcli/bsp2svg/converter.py:66]: nothing groundbreaking 😅 but it does look prettier when opened with Inkscape, in the "Layers and Objects" section
image

File: qcli/bsp2svg/converter.py (hunk 6 above)
Comment (alexkiri) [reply qcli/bsp2svg/converter.py:68]: not removed, but moved because of the reason mentioned above

File: qcli/bsp2svg/converter.py (hunk 5 above)
Comment (alexkiri) [reply qcli/bsp2svg/converter.py:39]: it was hardcoded in the first place, in the original code at line 74. I just moved the code and used a lambda to make it more concise. the reason for moving is because the faces need to be filtered _before_ calculating the min / max for the drawing dimensions

File: qcli/bsp2svg/converter.py (hunk 1 above)
Comment (alexkiri) [reply qcli/bsp2svg/converter.py:42]: ah good point. this part will help with the next PR about slicing, where the `faces` variable will be filtered again based on their `plane`. it's not necessary, but it will help with the future PR

File: qcli/bsp2svg/converter.py (hunk 7 above)
Comment (alexkiri) [reply qcli/bsp2svg/converter.py]: will do

File: qcli/bsp2svg/cli.py (hunk 4 above)
Comment (alexkiri) [reply qcli/bsp2svg/cli.py:82]: thanks!
I just had another idea. the ability to specify a negative axis, in order to flip the projection. this makes more sense for the x and y axis. what do you think? should I add this? how should I put the choices? something like `choices=['x', 'y', 'z', 'nx', 'ny', 'nz']`?

File: qcli/bsp2svg/converter.py (hunk 2 above)
Comment (joshuaskelly) [reply qcli/bsp2svg/converter.py:66]: Haha. Good enough for me!

File: qcli/bsp2svg/cli.py (hunk 4 above)
Comment (joshuaskelly) [reply qcli/bsp2svg/cli.py:82]: Maybe hold off on this for the moment? You can also accomplish this by manually flipping the svg drawing, correct?

File: qcli/bsp2svg/converter.py (hunk 6 above)
Comment (joshuaskelly) [reply qcli/bsp2svg/converter.py:68]: Ah I see. Thanks for clarifying

File: qcli/bsp2svg/converter.py (hunk 5 above)
Comment (joshuaskelly) [reply qcli/bsp2svg/converter.py:39]: So I did hard code it myself. 🤦‍♂️

File: qcli/bsp2svg/converter.py (hunk 1 above)
Comment (joshuaskelly) [reply qcli/bsp2svg/converter.py:42]: Gotcha. I understand better now. I'm totally fine with the `faces` list. 👍

File: qcli/bsp2svg/converter.py (hunk 7 above)
Comment (joshuaskelly) [reply qcli/bsp2svg/converter.py]: You made this change. Should we call this done?

File: qcli/bsp2svg/cli.py (hunk 4 above)
Comment (alexkiri) [reply qcli/bsp2svg/cli.py:82]: almost. the outer edge could look same when flipping the svg, but the content inside could be different. but yes, we can hold this off

File: qcli/bsp2svg/converter.py (hunk 7 above)
Comment (alexkiri) [reply qcli/bsp2svg/converter.py]: yup

File: qcli/bsp2svg/converter.py (hunk 5 above)
Comment (alexkiri) [reply qcli/bsp2svg/converter.py:39]: it works pretty well though 😄 and I think that the `sky` as part of the texture is actually defined in the game logic to be treated as a special texture. it's fine as it is
